[FLINK-10461] [State Backends, Checkpointing] Speed up download files when restore from DFS

klion26 opened the pull request:
## What is the purpose of the change

This PR speed up the download files when restore from DFS using multi-thread.

## Brief change log
Use multi-thread from download files when restore from DFS.

## Verifying this change

Add a UT for test exception rethrow in thread pool

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no)
  - If yes, how is the feature documented? (**not applicable**)

klion26 commented:
cc @StefanRRichter 
I  use 5 threads for downloading files, because I think 5 is enough for  a lot of scenarios, and from the experience of my company using 5 threads looks good.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java:
```diff
@@ -1286,10 +1289,38 @@ private void transferAllDataFromStateHandles(
 			Map<StateHandleID, StreamStateHandle> stateHandleMap,
 			Path restoreInstancePath) throws IOException {
 
+			ExecutorService executorService = Executors.newFixedThreadPool(5);
```
Number of threads should be configurable. For example, the same way as incremental checkpointing. At least in RocksDBStateBackend constructor and CheckpointingOptions. The default behaviour (e.g. number of threads 1) can be as it is now, basically running in current thread.

azagrebin replied:
Cancelation of `copyStateDataHandleData` futures and `executorService.shutdownNow()` should be registered in `cancelStreamRegistry.registerCloseable()` and unregistered with `unregisterCloseable` before `executorService.shutdown()` in `finally`. Similar to streams in `copyStateDataHandleData`. This way we guarantee freeing of all resources in case of job abrupt shutdown.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java:
```diff
@@ -1286,10 +1289,38 @@ private void transferAllDataFromStateHandles(
 				StreamStateHandle remoteFileHandle = entry.getValue();
-				copyStateDataHandleData(new Path(restoreInstancePath, stateHandleID.toString()), remoteFileHandle);
+
+				FutureTask<Void> task = new FutureTask<>(() -> {
```
I think it would be more maintainable if we extract these 3 methods:
- `transferAllStateDataToDirectory`
- `transferAllDataFromStateHandles`
- `copyStateDataHandleData`
into a separate class file, e.g. `RocksDbStateDataTransfer`.
`RocksDBKeyedStateBackend` already contains a lot.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java:
```diff
@@ -1286,10 +1289,38 @@ private void transferAllDataFromStateHandles(
+					return null;
+				});
+				tasks.add(task);
+				executorService.submit(task);
```
I suggest to use `CompletableFuture.runAsync(Runnable, Executor)` instead of `FutureTask`. It allows to use `FutureUtils.waitForAll(CompletableFutures).get()`.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackendTest.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Tests for RocksDBKeyedStateBackend.
+ */
+public class RocksDBKeyedStateBackendTest {
```
In addition to failure scenario, I would also then add a test which mocks `StreamStateHandle` streams and checks that all streams are read and written correctly in parallel.

klion26 commented:
Thank you for you review, @azagrebin  ! I will push another commit to address your comments.

klion26 commented:
Hi, @azagrebin , could you please help to review this?

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -235,9 +238,10 @@ public RocksDBStateBackend(StateBackend checkpointStreamBackend) {
 	 * @param enableIncrementalCheckpointing True if incremental checkpointing is enabled.
 	 */
-	public RocksDBStateBackend(StateBackend checkpointStreamBackend, TernaryBoolean enableIncrementalCheckpointing) {
+	public RocksDBStateBackend(StateBackend checkpointStreamBackend, TernaryBoolean enableIncrementalCheckpointing, int restoreThreadNum) {
```
I would also leave the previous method without restoreThreadNum:
```
public RocksDBStateBackend(StateBackend checkpointStreamBackend, TernaryBoolean enableIncrementalCheckpointing) {
    this(checkpointStreamBackend, enableIncrementalCheckpointing, -1);
}
```
this way we do not break existing code, including other methods here, like this modification which is not needed then and other similar ones:
```
public RocksDBStateBackend(StateBackend checkpointStreamBackend) {
		this(checkpointStreamBackend, TernaryBoolean.UNDEFINED, 1);
	}
```

and I would rather keep it undefined here, e.g. `-1`, similar to `TernaryBoolean.UNDEFINED` for `enableIncrementalCheckpointing`. It can be resolved then in `private RocksDBStateBackend(RocksDBStateBackend original, Configuration config)` and `getRestoreThreadNum`.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,149 @@
+	public static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int resotreThreadNum,
```
typo: `resotreThreadNum` -> `restoredThreadNum`, also in other methods

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,149 @@
+		List<Closeable> closeables = new LinkedList<>();
+
+		try {
+			closeables.add(() -> executorService.shutdownNow());
```
I think the tasks should be first properly canceled and then `executorService` should shutdown.
`executorService.shutdownNow` also calls `future.cancel` which is interrupting the task thread.
At the moment `while (true)` in `copyStateDataHandleData` does not support explicit cancelation.
I suggest to create an `running = AtomicBoolean(true)` in the beginning then `while (true)` can be `while(running.get())`. This will allow to cancel all tasks at once and shutdown `executorService` in one `closeable`:
`closeables.add(() -> { running.set(false); executorService.shutdownNow()); };`
List of `closeables` is not needed then.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,149 @@
+					CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
+						try {
+							copyStateDataHandleData(new Path(restoreInstancePath, stateHandleID.toString()), remoteFileHandle, closeableRegistry);
+						} catch (IOException e) {
```
the exception handling can go inside `copyStateDataHandleData`.
`new Path(restoreInstancePath, stateHandleID.toString())` can be a variable to reduce the line length.
I would create a list of `Runnable`'s:
```
Path path = new Path(restoreInstancePath, stateHandleID.toString());
runnables.add( -> copyStateDataHandleData(path, remoteFileHandle, closeableRegistry));
```
and then have separate code for `resotreThreadNum > 1` and `resotreThreadNum == 1`.
`executorService` currently is not used for `resotreThreadNum == 1` but created.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -0,0 +1,143 @@
+	public void testMultiThreadRestoreCorrectly() throws Exception {
+		IncrementalKeyedStateHandle stateHandle = mock(IncrementalKeyedStateHandle.class);
+
+		byte[] content1 = new byte[1];
```
I would create a list of something like 6 test contexts and use loops over them for further actions.
content length can be also random.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -0,0 +1,143 @@
+		while (b > -1) {
+			assertEquals(expected[idx++], b < 128 ? b : b - 256);
+			b = inputStream.read();
+		}
```
`Files.readAllBytes` can be probably used and then `assertArrayEquals`.

klion26 commented:
Hi, @azagrebin , I've just addressed all comments, please help to review this when you have time.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -236,8 +239,25 @@ public RocksDBStateBackend(StateBackend checkpointStreamBackend) {
+	public RocksDBStateBackend(StateBackend checkpointStreamBackend, TernaryBoolean enableIncrementalCheckpointing, int restoredThreadNum) {
 		this.checkpointStreamBackend = checkNotNull(checkpointStreamBackend);
 		this.enableIncrementalCheckpointing = enableIncrementalCheckpointing;
+		this.restoredThreadNum = restoredThreadNum;
```
Actually, instead of changing and exploding the constructor signatures, I would suggest to set it here to -1 (undefined). Then we could remove `final` from `restoredThreadNum` and add setter method for it, like for `predefinedOptions`. Users could use the setter to set this option per job.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -276,6 +312,12 @@ private RocksDBStateBackend(RocksDBStateBackend original, Configuration config)
 		this.enableIncrementalCheckpointing = original.enableIncrementalCheckpointing.resolveUndefined(
 			config.getBoolean(CheckpointingOptions.INCREMENTAL_CHECKPOINTS));
 
+		if (original.restoredThreadNum == CheckpointingOptions.CHECKPOINT_RESTORE_THREAD_NUM.defaultValue()) {
```
here we could initialise `restoredThreadNum` from `config` if `original.restoredThreadNum` is -1 (undefined). Otherwise `restoredThreadNum` = `original.restoredThreadNum`, similar to `original.enableIncrementalCheckpointing.resolveUndefined`.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -686,6 +729,13 @@ public RocksDBNativeMetricOptions getMemoryWatcherOptions() {
+	 * Gets the thread number will used for downloading files from DFS when restore.
+	 */
+	public int getRestoredThreadNum() {
+		return restoredThreadNum;
```
here we could return `restoredThreadNum` if it is not -1 (defined) and `CheckpointingOptions.CHECKPOINT_RESTORE_THREAD_NUM.defaultValue()` if `restoredThreadNum` is still -1 (undefined).

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,162 @@
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void transferAllDataFromStateHandles(
```
I think this methid should be broken down into 3 methods:
- loop which creates `List<Runnable>`
- multi-threaded version where `executorService`, `running` and `shutDownAllTasksAndExecutorService` are used and Runnables are submitted in a loop with try/catch.
- single-threaded version which just runs Runnables.
If `restoredThreadNum` > 1 then multi-threaded version should be called in this method else single-threaded version.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,162 @@
+				Executors.newFixedThreadPool(restoredThreadNum) :
+				null;
+
+		running.set(true);
```
`AtomicBoolean running = new AtomicBoolean(true);` 
can be local variable passed to `copyStateDataHandleData`

klion26 commented:
@azagrebin thanks for your review, I've just addressed all your comments and created[ FLINK-11008](https://issues.apache.org/jira/browse/FLINK-11008) for follow up issue. Please help to review this when you have time.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,185 @@
+		CloseableRegistry closeableRegistry,
+		AtomicBoolean running) throws ExecutionException, InterruptedException, IOException {
+
+		final ExecutorService executorService = enableMultiThreadDownload(restoredThreadNum) ?
```
Here we can already assume that `enableMultiThreadDownload  == true`, right? So it can be just:
`final ExecutorService executorService = Executors.newFixedThreadPool(restoredThreadNum);`

I also think it can be just `threadNum` here because it can be used also for upload same way.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,185 @@
+		List<Runnable> runnables = getRunnablesForStateHandles(stateHandleMap, restoreInstancePath, closeableRegistry, running);
+
+		if (enableMultiThreadDownload(restoredThreadNum)) {
+			runRunnalbesMultithread(runnables, restoredThreadNum, closeableRegistry, running);
```
I would rename like this:
transferAllDataFromStateHandles -> downloadDataForAllStateHandles
getRunnablesForStateHandles ->createDownloadRunnables
runRunnalbesMultithread  -> runTransferWithMultipleThreads
runRunnablesSingleThread  -> runTransferWithSingleThread
copyStateDataHandleData -> downloadDataForStateHandle

also I would rename everywhere:
restoredThreadNum -> restoringThreadNum

klion26 commented:
@azagrebin thanks for your reviewing, i'vs just addressed the comments, please help to review this when you have time.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOptions.java:
```diff
@@ -45,4 +45,12 @@
 			HEAP.name(), ROCKSDB.name()));
+
+	/**
+	 * The thread numbers used to download files from DFS in RocksDBStateBackend.
```
Also here without s:
`The thread number used`

klion26 commented:
Thanks a lot for your reviews, I learnt a lot from them. @azagrebin 
I fixed the java doc, and thank you again.

azagrebin commented:
Thanks for addressing the comment.

One more thing, I think we do not actually need the `running` flag because we register streams in closable registry. In case of canceling the job, the runnables will either fail to register streams or will be interrupted because streams will be closed unexpectedly and break the while loop anyways with IO exception which is what we want. The while loops will not hold executor shutdown for long. Sorry for confusion.

Can we still change it and remove the atomic `running`?

klion26 commented:
@azagrebin thank for your reply, I have a question about this, will JM failover lead to two download threads? If this is no, I'll change and remove the `running` flag, thanks.

azagrebin commented:
In general, I do not see any problem in having multiple threads, downloading even the same data. If your question is whether they can interfere with each other locally, the target directories are randomised so they should download to different locations, even if at the same time. Eventually, it is the JM job that only one of them will be used.

klion26 commented:
@azagrebin got it, thank you for the explanation, i've just remove the flag `running`

tillrohrmann commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -0,0 +1,122 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
```
Is it possible to write this test without mocking?

tillrohrmann commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,177 @@
+			runTransferWithMultipleThreads(runnables, restoringThreadNum, closeableRegistry);
+		} else {
+			runTransferWithSingleThread(runnables);
+		}
```
Wouldn't it be better to first instantiate the `ExecutorService` and then use it to download the files.
```
ExecutorService executorService = createExecutorService(restoringThreadNum);

for (Runnable runnable: createDownloadRunnables()) {
    executorService.execute(runnable);
}
```

if `restoringThreadNum == 1`, then we create a direct executor service. That way we would not need to duplicate the logic in `runTransferWithSingleThread`.

tillrohrmann replied:
`DirectExecutorService` would execute everything in the calling thread.

azagrebin replied:
I agree, we will have less code by just using the `DirectExecutorService` instead of looping Runnable's. Then code in `runTransferWithMultipleThreads` could go back to `downloadDataForAllStateHandles` and `runTransferWithSingleThread` would not be needed anymore.

klion26 replied:
I agree. I found `DirectExecutorService` in test code(`flink-runtime/src/test/java/org/apache/runtime/util/DirectExecutorService`), Do I miss something? or should I move it to `flink-runtime/src/main/java/org/apache/runtime/util/DirectExecutorService` ?

tillrohrmann commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,177 @@
+				executorService.shutdownNow();
+			}
+		};
+		closeableRegistry.registerCloseable(shutDownAllTasksAndExecutorService);
```
Why do we register the `ExecutorService` at the `CloseableRegistry`?

tillrohrmann replied:
I think this should not be needed.

azagrebin replied:
True, if the main thread, restoring from checkpoint, gets interrupted correctly, it will interrupt `FutureUtils.waitForAll` and executorService can be always `shutdownNow` in `finally` block.

tillrohrmann commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,177 @@
+				outputStream.write(buffer, 0, numBytes);
+			}
+		} catch (IOException e) {
+			throw new RuntimeException("Copy state from DFS error.", e);
```
Throwing `RuntimeExceptions` is discouraged. Please add the `throws IOException` clause and use a `ThrowingRunnable.unchecked` when giving it to the `Executor`.

tillrohrmann commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -686,6 +697,17 @@ public RocksDBNativeMetricOptions getMemoryWatcherOptions() {
+	 * Gets the thread number will used for downloading files from DFS when restore.
+	 */
+	public int getNumberOfRestoringThreads() {
+		return restoringThreadNum == -1 ? RocksDBOptions.CHECKPOINT_RESTORE_THREAD_NUM.defaultValue() : restoringThreadNum;
```
Can we also rename the field that it matches the getter?

azagrebin replied:
and `setrestoringThreadNum` as well then

tillrohrmann commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,172 @@
+				CompletableFuture<Void> future = CompletableFuture.runAsync(runnable, executorService);
+				futures.add(future);
+			}
+			FutureUtils.waitForAll(futures).get();
```
Let's remove the `ExecutionException` by unwrapping it via `ExceptionUtils.stripExecutionException`

klion26 commented:
@azagrebin Hi, For unifing the `threadNum > 1` and `threadNum = 1`, we should use `DirectExecutorService` for `threadNum = 1`, But I found `DirectExecutorService` locates in `flink-runtime/src/test/java/org/apache/runtime/util/DirectExecutorService` such as I commented above.

For this, I have a question, did I miss something? or should I move `flink-runtime/src/test/java/org/apache/runtime/util/DirectExecutorService` to `flink-runtime/src/main/java/org/apache/runtime/util/DirectExecutorService`

azagrebin commented:
@klion26 right, sorry for confusion.
I think we should move `DirectExecutorService` into `org.apache.flink.runtime.concurrent`.
We have already there `Executors.directExecutor()`. This could be deduplicated.
`directExecutor()` could become `directExecutorService()` and return `DirectExecutorService` singleton as `ExecutorService` instead of `Executor`. Then `DirectExecutor` is not needed in `Executors`.

Could you do it as a separate commit?

klion26 commented:
@azagrebin Thank you for your explanation. 
As your explanation, I will 
 - first move `DirectExecutorService` into `org.apache.flink.runtime.concurrent` and change the `Executor Executors#directExecutor()` to `DirectExecutorService Executors#direcExecutorService()`, this will be done in a separate commit.
- then use `Executors#direcExecutorSerivice()` in current patch to share the logic of `threadNum = 1` and `threadNum > 1`.

Did I understand right? If yes, I will file a issue and implement the first move, and then come back to complete this patch.

azagrebin commented:
@klion26, looks good, I think just `Executors#direcExecutorService()` can return `ExecutorService` interface. You could also just do it right here in this PR as a separate commit without a separate issue.

klion26 commented:
@azagrebin thank you for your reply. I've just addressed the comments.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/Executors.java:
```diff
@@ -59,6 +71,219 @@ public void execute(@Nonnull Runnable command) {
 		}
 	}
 
+	public static ExecutorService directExecutorService() {
```
I think we do not need the above `directExecutor()` method and `DirectExecutor` class. `directExecutorService()` can be used instead of `directExecutor()`.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/Executors.java:
```diff
@@ -59,6 +71,219 @@ public void execute(@Nonnull Runnable command) {
+	/**
+	 * Direct executor service implementation.
+	 */
+	private static class DirectExecutorService implements ExecutorService {
```
Let's move to separate class file. The class can be package private.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java:
```diff
@@ -264,6 +268,7 @@ public RocksDBKeyedStateBackend(
 		this.operatorIdentifier = Preconditions.checkNotNull(operatorIdentifier);
 
 		this.enableIncrementalCheckpointing = enableIncrementalCheckpointing;
+		this.restoringThreadNum = restoringThreadNum;
```
let's also name it `numberOfRestoringThreads`.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,145 @@
+			futures.add(CompletableFuture.runAsync(runnable, executorService));
+		}
+
+		FutureUtils.waitForAll(futures).get();
```
we still have to call `executorService.shutdownNow()` at the end:
```
try {
  List<Runnable > runnables = ...;
  ....
  FutureUtils.waitForAll(futures).get();
} finally {
  executorService.shutdownNow();
}
```

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	private static ExecutorService createExecutorService(int threadNum) {
+		if (threadNum == 1) {
```
I would rewrite:
```
if (threadNum > 1) {
  return Executors.newFixedThreadPool(threadNum);
} else {
  return directExecutorService();
}
```

klion26 commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -0,0 +1,158 @@
+		} catch (ExecutionException e) {
+			final Throwable throwable = ExceptionUtils.stripExecutionException(e);
+			if (throwable instanceof IOException) {
+				throw (IOException) throwable;
```
I followed [AbstractTaskManagerFileHandle](https://github.com/apache/flink/blob/951ae9a813cbf8a2db95519853f940277e4ea088/flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/AbstractTaskManagerFileHandler.java#L127) and [JobClient](https://github.com/apache/flink/blob/951ae9a813cbf8a2db95519853f940277e4ea088/flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java#L445) to throw `IOException` after `ExceptionUtils.stripExecutionException(e)` here, please correct me if have to change the behavior.

azagrebin replied:
I do not think we need to wrap non-IOException with IOException. I would leave it just as `throw ExceptionUtils.stripExecutionException(e)`

klion26 replied:
Thank you for the explanation @azagrebin.
I found `throw ExceptionUtils.stripExecutionException(e)` will throw `Throwable`, so we should add `throws Throwable` in the method signature.
How about use the blow code
```
try {
...
} catch (ExecutionException e) {
    final Throwable throwable = ExceptionUtils.stripExecutionException(e);
    ExceptionUtils.rethrow(throwable);
} finally {
    executorService.shutdownNow();
}
```
Because I found the `ExecutionException` will be thrown when executing the runnalbes, and the runnables were wrapped by `ThrowingRunnable.unchecked`, `ThrowingRunnable.unchecked` will only rethrow `Error` or `RuntimeExeception` by `ExeceptionUtils.rethrow`.

klion26 commented:
thank you @azagrebin .I've just addressed the comments.

zentol commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -0,0 +1,151 @@
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for RocksDbStateDataTransfer.
```
use `@link` for referencing `RocksDbStateDataTransfer`

zentol commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -0,0 +1,151 @@
+ */
+public class RocksDBStateDataTransferTest extends TestLogger {
+	@Rule
+	public TemporaryFolder temporaryFolder = new TemporaryFolder();
```
can be final

klion26 commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -686,6 +697,17 @@ public RocksDBNativeMetricOptions getMemoryWatcherOptions() {
+	 * Gets the thread number will used for downloading files from DFS when restore.
+	 */
+	public int getNumberOfRestoringThreads() {
+		return numberOfRestoringThreads == -1 ? RocksDBOptions.CHECKPOINT_RESTORE_THREAD_NUM.defaultValue() : numberOfRestoringThreads;
```
@zentol thank you for your review. How about add a variable `private final int undefinedThreadNum = -1` in  `RocksDBStateBackend`, change the code in constructor to `this.numberOfRestoringThreads = undefinedThreadNum` and here `return numberOfRestoringThreads == undefinedThreadNum ? RocksDBOptions.CHECKPOINT_RESTORE_THREAD_NUM.defaultValue() : numberOfRestoringThreads;`

azagrebin replied:
Let's keep it in sync with the logic for `enableIncrementalCheckpointing` for now. 
Later, we can revisit/refactor the backend configuration.

zentol commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -120,6 +120,9 @@
 	private final TernaryBoolean enableIncrementalCheckpointing;
 
+	/** Thread number used to download from DFS when restore, default value: 1. */
+	private int numberOfRestoringThreads;
```
i think this can also be final

azagrebin replied:
We introduced a setter for it: `setNumberOfRestoringThreads`, not to explode the number of constructor args like for `enableIncrementalCheckpointing`. At the moment `RocksDBStateBackend` basically serves as a configuration object and factory for the keyed and operator backends.

klion26 commented:
@azagrebin @zentol thank you for the review, I've just updated the code, leave the logic of `getNumberOfRestoringThreads ` unchanged, in sync with the logic for `enableIncrementalCheckpointing` for now.
